src/main/java/bio/terra/cli/command/notebooks/Create.java

```diff
@@ -92,11 +101,57 @@ protected void execute() {
     envVars.put("SUBNET", "projects/" + projectId + "/regions/" + zone + "/subnetworks/subnetwork");
     envVars.put("TERRA_WORKSPACE_ID", workspaceContext.getWorkspaceId().toString());
 
+    OUT.println(
+        String.format(
+            "Creating notebook instance `projects/%s/locations/%s/instanceId/%s`",
+            projectId, location, instanceName));
     new DockerAppsRunner(globalContext, workspaceContext).runToolCommand(command, envVars);
 
     OUT.println(
         "Notebook instance starting. This will take ~5-10 minutes.\n"
             + "See your notebooks in this workspace at https://console.cloud.google.com/ai-platform/notebooks/list/instances?project="
             + projectId);
   }
+
+  /**
+   * Returns the specified instanceName or an auto generated instance name with the username and
+   * date time.
+   */
+  // TODO add some unit tests when we have a testing framework.
+  private String getInstanceName(TerraUser user) {
+    if (!AUTO_GENERATE_NAME.equals(rawInstanceName)) {
+      return rawInstanceName;
+    }
+    String mangledUsername = mangleUsername(extractUsername(user.terraUserEmail));
+    String localDateTimeSuffix =
+        DateTimeFormatter.ofPattern(AUTO_NAME_DATE_FORMAT)
+            .format(Instant.now().atZone(ZoneId.systemDefault()));
+    return mangledUsername + localDateTimeSuffix;
+  }
+
+  /**
+   * Best effort mangle the user's name so that it meets the requirements for a valid instance name.
+   *
+   * <p>Instance name id must match the regex '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)'. I don't have
+   * a documentation link, but gcloud will complain otherwise.
+   */
+  private static String mangleUsername(String username) {
+    // Strip non alpha-numeric or '-' characters.
+    String mangledName = username.replaceAll("[^a-zA-Z0-9-]", "");
+    if (mangledName.isEmpty()) {
+      mangledName = "notebook";
+    }
+    // Lower case everything, even though only the first character requires lowercase.
+    mangledName = mangledName.toLowerCase();
+    // Make sure the returned name isn't too long to not have the date time suffix.
+    int maxNameLength = 61 - AUTO_NAME_DATE_FORMAT.length();
```

Comment:
nit: Where does 61 come from? I think defining that as a static/constant with a descriptive name would be helpful.

Reply:
good point, done.